### PR TITLE
Refactor command groupings for help in kubectl

### PR DESCRIPTION
**Pull request by pwittrock**
Rearrange the kubectl grouping commands.

LMK what you think.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35657)

### CODE REVIEW

**Comment by fabianofranz** [comment]
@kubernetes/kubectl

File: pkg/kubectl/cmd/cmd.go [hunk 1]
```diff
@@ -219,69 +219,64 @@ func NewKubectlCommand(f cmdutil.Factory, in io.Reader, out, err io.Writer) *cob
...
-				rollout.NewCmdRollout(f, out),
+				NewCmdGet(f, out, err),
 				NewCmdRollingUpdate(f, out),
-				NewCmdScale(f, out),
-				NewCmdAutoscale(f, out),
 			},
 		},
 		{
-			Message: "Cluster Management Commands:",
+			Message: "Application Management (Using Yaml Config)",
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/cmd.go]: I don't know that most new users have any idea what YAML is, so this has the potential for confusion.  "Application Management (via config files)" might be better.

**Comment by smarterclayton** [comment]
Ordering is ok, I'm weirded out by saying "Command args / flags" or "Yaml".  Maybe talk about "config files" vs "direct changes", but even that is very unusual.  I think most people approaching kube the first time think of "direct changes or command args" as the only way things can be done, so config management is novel (or at least, separate).

I think it's more important for new users to find create / delete without confusion than it is to talk about how they can be config driven.  And since apply handles create as well, create is really more direct action and doesn't need a label per se.

File: pkg/kubectl/cmd/cmd.go [hunk 2]
```diff
@@ -219,69 +219,64 @@ func NewKubectlCommand(f cmdutil.Factory, in io.Reader, out, err io.Writer) *cob
...
+				NewCmdLabel(f, out),
+				rollout.NewCmdRollout(f, out),
 				NewCmdRun(f, in, out, err),
+				NewCmdScale(f, out),
 				set.NewCmdSet(f, out, err),
 			},
 		},
 		{
-			Message: "Basic Commands (Intermediate):",
+			Message: "Application Management (Either Yaml Or Command Args)",
```
Comment (AdoHe) [inline pkg/kubectl/cmd/cmd.go]: `Either Config Files Or Command Args`, wdyt?

Comment (smarterclayton) [inline pkg/kubectl/cmd/cmd.go]: That's still really verbose - as a user, I'm not sure that helps me.  The concept we're trying to convey is "Use these commands to manage your applications directly" which is slightly different than "Use these commands to treat your applications as config" or "manage applications using config".

**Comment by pwittrock** [comment]
WDYT of 

>
Command Driven App Management
Config Driven App Management

or

>
App Management Commands
App Config Management

Also, I torn on what do to with commands like "Create" that have direct management sub-commands but the top level command is config based?  Or rollingupdate that is driven by either config or flags for the same command?  Thoughts?

**Comment by pwittrock** [comment]
@smarterclayton Do you have any suggestions for how to better group the commands?  Since most users will be coming from the direct management world initially, how would you recommend we group all of the direct management commands together?

**Comment by smarterclayton** [comment]
"App Config Management" is pretty good - I think that's a good balance between the things you're trying to convey.

The `create` vs `create TYPE` separation sucks.  I almost think we should try to figure out how to cleanly show those in two different groups (@fabianofranz) in the long term, although in the short term that might be complicated.  

For direct management I think it makes sense to have "initial experience" commands (that focus on things users *must* know immediately, no more than 3-4), and then "app management" commands (that are the direct management), then the more complex app management commands.    However, most users are going to start with create so it's probably ok to put "create" in the "initial experience" commands and just punt on the split for now.  

Ultimately the help is for a new user trying to find something, rather than experienced users, so we should bias towards classic information architecture choices (split into sets of concepts, have easiest concepts up front, use groups that correspond to users skill sets), etc.

**Comment by pwittrock** [comment]
Makes sense to me.

The --help is the closest thing we have to reference documentation, and so I have been thinking about this from that perspective.

IMHO there are a couple different problems kubectl tries to solve - most clearly the cluster management vs app management commands.  To me the `command based app mg` and `config based app ` commands seem like they could have even been put in separate cli tools.  The groups I have been creating have been from trying to break up the commands into what problem they are trying to solve and generally the type of solution provided.

Subdividing the groups into initial exp vs more complex makes sense from a discovery perspective but IMHO provides a less clear distinction of "what are these commands" vs "what are those commands".

Alternatively, we could order the commands in each group by how experienced the user would be before they would likely encounter the need for this command.  This makes it less clear what the organization is vs alphanum ordering, but I think offers a decent balance.

WDYT?

**Comment by fabianofranz** [comment]
> The create vs create TYPE separation sucks. I almost think we should try to figure out how to cleanly show those in two different groups

Not too hard, it's doable with a different help template specifically created for when invoking help with `kubectl -h`, or even better, an attribute in the one help template that would be set by the help func only in the top-level cmd.

```
App Management Commands:
  create quota    Create a quota with the specified name
  create secret   Create a secret using specified subcommand
  (...)
App Config Management:
  create          Create a resource by filename or stdin
  (...)
```

I kind of like it. 

[spinoff] Also makes me think that a command/help for having an overview of the entire cmd tree (like a "command sitemap") would be interesting for findability, discoverability for not-exactly-first-time users, etc.[/spinoff]

**Comment by AdoHe** [comment]
```
command sitemap
```
sounds interesting :)

**Comment by pwittrock** [comment]
This SGTM.  I am not sure what the policy is with this sort of thing and feature freeze.  I am traveling for the next 2 weeks, so I may not be able to wrap this up in time for 1.5.

**Comment by AdoHe** [comment]
@pwittrock any special reason to close this? I think we still need to refactor the command group.

**Comment by pwittrock** [comment]
No, I can leave it open.  I won't have time to work on this for a month or so and it was creating noise in my PR queue list.

**Pull request closed without merging**